### ISSUE TITLE
ninja rebuilds lean.exe after every Git commit on Windows

### ISSUE THREAD

**Issue by htzh** [leanprover/lean3#733]
For some reason on Windows ninja starts with rerunning CMake and always rebuild the binary

```
[1/4] Building CXX object shell/CMakeFiles/lean.dir/lean.cpp.obj
[2/4] Linking CXX executable shell\lean.exe
```

even when there is no change to C++ source files.
The consequence (I think) is that all lean files get recompiled instead of just those that have changed everytime I run ninja.

**Comment by soonhokong** [comment]
@htzh, could you give me the result of `ninja -d explain`?

**Comment by htzh** [comment]
I should revise the problem report to "ninja rebuilds binary **after a Git commit** even when no c++ source file has changed". If I rerun ninja right after a build it does not attempt to rebuild the binary or the library. I didn't run `ninja -d explain` after the commit so I don't have the result to report here. I will remember to run ninja with the flags next time I make a change to the library.

**Comment by soonhokong** [comment]
Can you show me the last part of your `build.ninja` file? There should be a line starting with `build build.ninja: RERUN_CMAKE`. I'm curious whether you have `.git/refs/heads/master` or not.

**Comment by htzh** [comment]
The file reads like:

```
#############################################
# Re-run CMake if any of its inputs changed.

build build.ninja: RERUN_CMAKE | C$:/Users/haitao/proj/lean/fork/.git/HEAD C$:/Users/haitao/
proj/lean/fork/.git/refs/heads/master C$:/Users/haitao/proj/lean/fork/src/../bin/leanemacs.b
at.in C$:/Users/haitao/proj/lean/fork/src/../bin/leanemacs.in C$:/Users/haitao/proj/lean/for
k/src/../bin/linja.in ......
```

The other files all appear to be CMake files.

**Comment by soonhokong** [comment]
I think this comes from the way CMake generates `build.ninja` file which is a recipe for `ninja` utility. As you see, it injected a dependency on `.git/HEAD` file. As a result, whenever it's updated, ninja is doing some tasks. As I see the portion of the file that you shared, it re-invoke CMake for sure. It's possible it triggers other builds as a result.

Is it simply linking lean binary again from the existing compiled objects or it compiles everything from scratch?

**Comment by htzh** [comment]
For lean.exe it just relinks the object files and rebuilds the executable. But that causes the library to be recompiled. For the library I think it rebuilds everything:

```
[1/96] "C:\Users\haitao\proj\lean\fork\bin\lean.exe"  \Users\haitao\proj\lean\fork\hott\init\datatypes.hlean -o "\Users\haitao\proj\lean\fork\hott\init\datatypes.olean" -c "\Users\haitao\proj\lean\fork\hott\init\datatypes.clean" -i "\Users\haitao\proj\lean\fork\hott\init\datatypes.ilean"
......
[1/163] "C:\Users\haitao\proj\lean\fork\bin\lean.exe"  \Users\haitao\proj\lean\fork\library\init\datatypes.lean -o "\Users\haitao\proj\lean\fork\library\init\datatypes.olean" -c "\Users\haitao\proj\lean\fork\library\init\datatypes.clean" -i "\Users\haitao\proj\lean\fork\library\init\datatypes.ilean"
......
```

**Comment by soonhokong** [comment]
I think this is what happens:
- Git commit updates `.git/HEAD`
- It triggers re-running CMake
- It re-generates files written by CMake (i.e. version.h)
- It triggers re-building `lean` and other binaries
- It triggers re-building lean libraries which depend on `lean`

It's because the build system is based on time-stamp, not its content. I don't think there is a good solution for this.

**Comment by htzh** [comment]
Yes I agree. It is odd that ninja/cmake depends on Git. Do they not work with other version control systems or when there is no version control? I know CMake works but I don't know ninja in other context. Feel free to close this issue as this does not appear to be Lean specific.

**Comment by soonhokong** [comment]
I believe that this is actually a feature of CMake while I don't think of a reason why they have it. Of course, CMake doesn't require that source code is hosted in a git or any DVCS.

> Feel free to close this issue as this does not appear to be Lean specific.

OK. I'll close it for now. Please feel free to re-open it anytime.

**Comment by htzh** [comment]
As a reference this the most recent `ninja -d explain` after a git rebase:

```
ninja explain: output build.ninja older than most recent input C:/Users/haitao/proj/lean/fork/.git/HEAD (458784406 vs 458800376)
ninja explain: output shell/CMakeFiles/lean.dir/lean.cpp.obj older than most recent input githash.h (458784408 vs 458801013)
ninja explain: shell/CMakeFiles/lean.dir/lean.cpp.obj is dirty
ninja explain: shell/lean.exe is dirty
ninja explain: CMakeFiles/hott_lib is dirty
ninja explain: shell/lean.exe is dirty
ninja explain: hott_lib is dirty
ninja explain: shell/lean.exe is dirty
ninja explain: CMakeFiles/standard_lib is dirty
ninja explain: shell/lean.exe is dirty
ninja explain: standard_lib is dirty
ninja explain: shell/lean.exe is dirty
```

Is there a way to remove the dependence on `githash.h`? I think using githash to track version causes a unnecessary recompile of the library. Is there another way to track lean.exe version based on timestamp from the lean src directory only?

**Comment by htzh** [comment]
@soonhokong I think this issue should be reopened as it appears that this is related to how Lean tracks build version, which seems to be something that can be addressed by Lean devs. I don't have permission to reopen issues.

**Comment by soonhokong** [comment]
@htzh, could you test https://github.com/soonhokong/lean/commit/324002b7badc515a4d5772344bfddaf49e88afd2 ? Please try to pass `-DUSE_GITHASH=OFF` when you run CMake.

**Comment by htzh** [comment]
I downloaded a fresh clone of your repo and run `cmake -DUSE_GITHASH=OFF -G Ninja ../src`. The build failed with

```
[372/379] Building CXX object tests/frontends/lean/CMakeFiles/lean_scanner.dir/scanner.cpp.obj
FAILED: C:\alt\msys64\mingw64\bin\c++.exe    -Wall -Wextra -std=c++11  -D LEAN_WINDOWS -D LEAN_WIN_STACK_SIZE=104857600 -D LEAN_MULTI_THREAD -D LEAN_CACHE_EXPRS -DLEAN_BUILD_TYPE="Release" -D LEAN_TRACK_MEMORY -I C:/alt/msys64/mingw64/include -D LEAN_USE_LUA_NEWSTATE -O3 -DNDEBUG -IC:/alt/msys64/mingw64/include -IC:/Users/haitao/proj/test/lean/src -I. -MMD -MT shell/CMakeFiles/lean.dir/lean.cpp.obj -MF shell/CMakeFiles/lean.dir/lean.cpp.obj.d -o shell/CMakeFiles/lean.dir/lean.cpp.obj -c C:/Users/haitao/proj/test/lean/src/shell/lean.cpp
C:/Users/haitao/proj/test/lean/src/shell/lean.cpp:45:31: fatal error: githash.h: No such file or directory
 #include "githash.h" // NOLINT
```

I think another possibility is to read in `githash.h` (if needed) as a data file instead of including it. It is kind of backwards to include it. The general workflow is edit/compile/test (I assume) and commit after it passes the test. Including `githash.h` would then always cause extra compile and build and the tested binary is different from the final binary so tests need to be rerun as well.

**Comment by soonhokong** [comment]
There was a mistake. Please could you test https://github.com/soonhokong/lean/commit/c72e661a9e81631825209264ccc84017198ec012 ?

**Comment by htzh** [comment]
The new commit works. I did a test commit and it no longer rebuilds the lean.exe and the whole library.

**Comment by soonhokong** [comment]
Great. I'll push the commit to the blessed repo and close this one.

**Comment by soonhokong** [comment]
Closed by c72e661.
